module/zfs/qat_compress.c

```diff
@@ -0,0 +1,528 @@
+}
+
+static void
+qat_clean(void)
+{
+	Cpa16U buff_num = 0;
+	Cpa16U num_inter_buff_lists = 0;
+	Cpa16U inst = 0;
+
+	for (inst = 0; inst < num_inst; inst++) {
+		cpaDcStopInstance(dc_inst_handles[inst]);
+		PHYS_CONTIG_FREE(session_handles[inst]);
+		/* free intermediate buffers  */
+		if (buffer_array[inst] != NULL) {
+			cpaDcGetNumIntermediateBuffers(
+			    dc_inst_handles[inst], &num_inter_buff_lists);
+			for (buff_num = 0; buff_num < num_inter_buff_lists;
+			    buff_num++) {
+				CpaBufferList *buffer_inter =
+				    buffer_array[inst][buff_num];
+				PHYS_CONTIG_FREE(buffer_inter->pBuffers->pData);
+				PHYS_CONTIG_FREE(buffer_inter->pBuffers);
+				PHYS_CONTIG_FREE(
+				    buffer_inter->pPrivateMetaData);
+				PHYS_CONTIG_FREE(buffer_inter);
+			}
+		}
+	}
+
+	num_inst = 0;
+	qat_init_done = 0;
+}
+
+int
+qat_init(void)
+{
+	CpaStatus status = CPA_STATUS_SUCCESS;
+	Cpa32U sess_size = 0;
+	Cpa32U ctx_size = 0;
+	Cpa16U num_inter_buff_lists = 0;
+	Cpa16U buff_num = 0;
+	Cpa32U buff_meta_size = 0;
+	CpaDcSessionSetupData sd = {0};
+	Cpa16U inst;
+
+	if (qat_init_done != 0)
+		return (0);
+
+	status = cpaDcGetNumInstances(&num_inst);
+	if (status != CPA_STATUS_SUCCESS || num_inst == 0)
+		return (-1);
+
+	if (num_inst > MAX_INSTANCES)
+		num_inst = MAX_INSTANCES;
+
+	status = cpaDcGetInstances(num_inst, &dc_inst_handles[0]);
+	if (status != CPA_STATUS_SUCCESS)
+		return (-1);
+
+	for (inst = 0; inst < num_inst; inst++) {
+		cpaDcSetAddressTranslation(dc_inst_handles[inst],
+		    (void*)virt_to_phys);
+
+		status = cpaDcBufferListGetMetaSize(dc_inst_handles[inst],
+		    1, &buff_meta_size);
+
+		if (status == CPA_STATUS_SUCCESS)
+			status = cpaDcGetNumIntermediateBuffers(
+			    dc_inst_handles[inst], &num_inter_buff_lists);
+
+		if (status == CPA_STATUS_SUCCESS && num_inter_buff_lists != 0)
+			status = PHYS_CONTIG_ALLOC(&buffer_array[inst],
+			    num_inter_buff_lists *
+			    sizeof (CpaBufferList *));
+
+		for (buff_num = 0; buff_num < num_inter_buff_lists;
+		    buff_num++) {
+			if (status == CPA_STATUS_SUCCESS)
+				status = PHYS_CONTIG_ALLOC(
+				    &buffer_array[inst][buff_num],
+				    sizeof (CpaBufferList));
+
+			if (status == CPA_STATUS_SUCCESS)
+				status = PHYS_CONTIG_ALLOC(
+				    &buffer_array[inst][buff_num]->
+				    pPrivateMetaData,
+				    buff_meta_size);
+
+			if (status == CPA_STATUS_SUCCESS)
+				status = PHYS_CONTIG_ALLOC(
+				    &buffer_array[inst][buff_num]->pBuffers,
+				    sizeof (CpaFlatBuffer));
+
+			if (status == CPA_STATUS_SUCCESS) {
+				/*
+				 *  implementation requires an intermediate
+				 *  buffer approximately twice the size of
+				 *  output buffer, which is 2x max buffer
+				 *  size here.
+				 */
+				status = PHYS_CONTIG_ALLOC(
+				    &buffer_array[inst][buff_num]->pBuffers->
+				    pData, 2 * QAT_MAX_BUF_SIZE);
+				buffer_array[inst][buff_num]->numBuffers = 1;
+				buffer_array[inst][buff_num]->pBuffers->
+				    dataLenInBytes = 2 * QAT_MAX_BUF_SIZE;
+			}
```

Comment:
for the above few lines of code, it doesn't handle the case memory allocation failure at all, though I don't know if ```cpaDcStartInstance()``` below can handle the case correctly. The code seems suspicious.